scripts/docker/docker.lib.sh

```diff
@@ -0,0 +1,256 @@
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+#   check=[output string to search for]
+function docker-check-test() {
+
+  local dir=${dir:-$PWD}
+  # shellcheck disable=SC2086,SC2154
+  docker run --rm --platform linux/amd64 \
+    ${args:-} \
+    "${DOCKER_IMAGE}:$(_get-version)" 2>/dev/null \
+    ${cmd:-} \
+  | grep -q "${check}" && echo PASS || echo FAIL
+}
+
+# Run Docker image.
+# Arguments (provided as environment variables):
+#   args=[arguments to pass to Docker to run the container, default is none/empty]
+#   cmd=[command to pass to the container for execution, default is none/empty]
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function docker-run() {
+
+  local dir=${dir:-$PWD}
+  # shellcheck disable=SC2086
+  docker run --rm --platform linux/amd64 \
+    ${args:-} \
+    "${DOCKER_IMAGE}:$(_get-version)" \
+    ${cmd:-}
+}
+
+# Push Docker image.
+# Arguments (provided as environment variables):
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function docker-push() {
+
+  local dir=${dir:-$PWD}
+  # Push all the image tags based on the stated versions
+  for version in $(_get-all-versions); do
+    docker push "${DOCKER_IMAGE}:${version}"
+  done
+  docker push "${DOCKER_IMAGE}:latest"
+}
+
+# Remove Docker resources.
+# Arguments (provided as environment variables):
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function docker-clean() {
+
+  local dir=${dir:-$PWD}
+  docker rmi "${DOCKER_IMAGE}:latest" > /dev/null 2>&1 ||:
+  for version in $(_get-all-versions); do
+    docker rmi "${DOCKER_IMAGE}:${version}" > /dev/null 2>&1 ||:
+  done
+  rm -f \
+    .version \
+    Dockerfile.effective
+}
+
+# ==============================================================================
+# Functions to be used with external images.
+
+# Retrieve the Docker image version from the '.tool-versions' file and pull the
+# image if required. This function is to be used in conjunction with the
+# external images and it prevents Docker from downloading an image each time it
+# is used, since the digest is not stored locally for compressed images. To
+# optimise, the solution is to pull the image using its digest and then tag it,
+# checking this tag for existence for any subsequent use.
+# Arguments (provided as environment variables):
+#   name=[full name of the Docker image]
+function docker-get-image-version-and-pull() {
+
+  # Get the image full version from the '.tool-versions' file
+  versions_file="$(git rev-parse --show-toplevel)/.tool-versions"
+  version="latest"
+  if [ -f "$versions_file" ]; then
+    line=$(grep "docker/${name} " "$versions_file" | sed "s/^#\s*//; s/\s*#.*$//")
+    [ -n "$line" ] && version=$(echo "$line" | awk '{print $2}')
+  fi
+
+  # Split the image version into two, tag name and digest sha256
+  tag="${version%%@*}"
+  digest="${version##*@}"
```

Comment:
This is my ignorance showing here. What are the `%%@*` and `##*@` suffixes doing?